### ISSUE TITLE
What is the difference between best and metrics

### ISSUE THREAD

**Issue by NingYuanLin** [ultralytics/yolov5#7334]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi.
I have trained yolov5 based on my datasets but I saw a puzzling summary in wandb: best and metrics
<img width="465" alt="image" src="https://user-images.githubusercontent.com/57001533/162216387-f82950a6-491a-467d-982f-43b544d3a084.png">
My question is:
1. Why `mAP_0.5` in metrics is higher than best?(as the showing of above picture)
2. Is the mAP calculated on the training set or the validation set?
3. I found the `metrics/mAP_0.5` reached 0.9312(as the showing of below picture) in chart at least but why it is different from `metircs` in summary
<img width="460" alt="image" src="https://user-images.githubusercontent.com/57001533/162217440-ffd502fb-2ecf-4b10-a4ab-5319150cdb4e.png">

**Comment by jyrr** [comment]
1. From my understanding;

The following validation run is ran on the end of training https://github.com/ultralytics/yolov5/blob/676e10cf1abc03360b56d8030adea2cd0d0af353/train.py#L443-L467

and differs a little from the validation run on the end of each epoch
https://github.com/ultralytics/yolov5/blob/5f941a84efdd45c986cd1c3764ced99e7c8e8294/train.py#L382-L404

resulting in slightly different metrics (but for the same model).

2. You can see that (in both cases) the validation data (val_loader) is used to calculate mAP.
3. This is because the model is not evaluated solely on mAP.5 but on a weighted combination of mAP.5 (*.1) & mAP.5-.95 (*.9)
https://github.com/ultralytics/yolov5/blob/5f941a84efdd45c986cd1c3764ced99e7c8e8294/utils/metrics.py#L15-L18

**Comment by AyushExel** [comment]
I think @jyrr's explanation is right. Maybe @glenn-jocher can confirm

**Comment by NingYuanLin** [comment]
Thank for @jyrr's detailed explanation and I think you are right. 
Thanks.

**Comment by mesllo-bc** [comment]
I still don't understand. When I independently rerun the validation for some model the metrics it outputs are not the same as the ones in the `best_metrics.json` that were saved during training. For example on `best_metrics.json` I have:

`{"precision": 0.7078748556693851, "recall": 0.546094897916718, "mAP_0.5": 0.6178649358384973`

while the validation logs show:

`all       3917      15852      0.683      0.549      0.598      0.447`

Shouldn't the precision, recall and mAP values check out here? The fitness function is not changing these values whatsoever it's only making a check and then saving the same results on `best_metrics.json`. What am I missing?

**Comment by glenn-jocher** [comment]
@mesllo-bc in your case, if the saved `best_metrics.json` does not match the validation logs, it could be due to multiple reasons such as different data splits, different random seeds or any post-processing applied during validation. Also, the computation of the mAP in the `metrics.py` may involve additional steps not performed during the regular validation run, leading to differences. Can you compare the `best_metrics.json` and validation logs to understand what could be causing the discrepancies? If you need further assistance, please provide more details about your setup.
